ast/impl/identifier_resolver.cpp

```diff
@@ -0,0 +1,410 @@
+#include "visitors.hpp"
+#include <memory>
+#include <set>
+
+namespace visitors {
+
+void IdentifierResolver::visit(Program* node) {
+    // First, verify that no 2 routines have the same name
+    for (auto routine : node->routines) {
+        if (routines.find(routine->name) != routines.end()) {
+            // TODO: Move the following to a function/macro
+            m_errors.push_back(Error{
+                .pos = routine->begin,
+                .message = "A routine already exists with that name",
+            });
+            return;
+        }
+        routines.insert({routine->name, routine});
+    }
+
+    // Then, verify that global variables names are unique (from routines too)
+    for (auto variable : node->variables) {
+        // Any routine with the same name?
+        if (routines.find(variable->name) != routines.end()) {
+            m_errors.push_back(Error{
+                .pos = variable->begin,
+                .message = "A routine already exists with that name",
+            });
+            return;
+        }
+        // Any other global variable with the same name?
+        for (auto definedVariable : variables) {
+            if (definedVariable->name == variable->name) {
+                m_errors.push_back(Error{
+                    .pos = variable->begin,
+                    .message = "Another variable with the same name exists",
+                });
+                return;
+            }
+        }
+        variables.push_back(variable);
+    }
+
+    for (auto typeDecl : node->types) {
+        types.push_back(typeDecl);
+    }
+
+    // Finally, visit all routine declaration
+    for (auto routine : node->routines) {
+        routine->accept(*this);
+    }
+
+    variables.clear();
+    routines.clear();
+    types.clear();
+}
+
+void IdentifierResolver::visit(RoutineDecl* node) {
+    // To keep track of the scope, each function that introduces a new scope
+    //  will append its declarations to the `variables` vector and then remove
+    //  them again before returning.
+    // When searching for the closest variable declaration later on, the same
+    //  vector will be iterated on from end to start
+
+    auto oldSize = variables.size();
+
+    for (auto parameter : node->parameters) {
+        variables.push_back(parameter);
+        parameter->type->accept(*this);
+    }
+
+    node->body->accept(*this);
+
+    variables.resize(oldSize);
+}
+
+void IdentifierResolver::visit(Type*) {}
+
+void IdentifierResolver::visit(AliasedType* node) {
+    for (auto it = types.rbegin(); it != types.rend(); it++) {
```

Comment:
Sorry, got that it is implemented in the body

Reply:
Yes, because the `visit(Body*)` overload pushes all types declared in the body. There was a minor problem in the order of pushing names, but I fixed it and will push now